### PR TITLE
Reject CD-ROM devices that use the virtio bus

### PR DESCRIPTION
**Pull request by stu-gott**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1802001

**Special notes for your reviewer**:

**Release note**:

```release-note
Default bus for CD-ROM devices is now sata.
```

### CODE REVIEW

**Comment by stu-gott** [comment]
/hold

Let's think through the side effects of forcing/not-forcing a default bus before merging.

On the one hand, creating a default bus other than virtio might be surprising. On the other, existing manifests that didn't specify a bus (that can happen, right?) will be invalid on accident.

**Comment by vladikr** [comment]
afaik, cdroms are not allowed on he virtio bus anymore: https://www.redhat.com/archives/libvir-list/2019-January/msg01104.html

**Comment by stu-gott** [comment]
/hold cancel

Ok, then this PR basically does the right thing.

**Comment by vladikr** [comment]
@stu-gott perhaps we should just reject `virtio`, I think cd-roms are allowed on `SCSI` or `ide`(not q35) ?

**Comment by stu-gott** [comment]
I don't understand. We do reject virtio requests for CD-ROM with this PR.

**Comment by stu-gott** [comment]
After testing I've confirmed that you're correct @vladikr. scsi bus should also be valid.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go [hunk 1]
```diff
@@ -2126,6 +2126,29 @@ var _ = Describe("Validating VMICreate Admitter", func() {
...
+			vmi.Spec.Volumes = append(vmi.Spec.Volumes, v1.Volume{
+				Name: "testcdrom",
+				VolumeSource: v1.VolumeSource{
+					ContainerDisk: &v1.ContainerDiskSource{Image: "fake"},
+				},
+			})
+
+			causes := ValidateVirtualMachineInstanceSpec(k8sfield.NewPath("fake"), &vmi.Spec, config)
+			Expect(len(causes)).To(Equal(1))
+			Expect(causes[0].Field).To(Equal("fake.domain.devices.disks[0].cdrom.bus"))
```
Comment (vladikr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go]: @stu-gott this test fails. It expects: 
`fake.domain.devices.disks.domain.devices.disks[0].cdrom.bus`

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 2]
```diff
@@ -1583,6 +1583,16 @@ func validateDisks(field *k8sfield.Path, disks []v1.Disk) []metav1.StatusCause {
 						Field:   field.Index(idx).Child(diskType, "bus").String(),
 					})
 				}
+
+				// special case. virtio is incompatible with CD-ROM for q35 machine types
+				if diskType == "cdrom" && bus == "virtio" {
+					causes = append(causes, metav1.StatusCause{
+						Type:    metav1.CauseTypeFieldValueInvalid,
+						Message: fmt.Sprintf("Bus type %s is invalid for CD-ROM device", bus),
+						Field:   field.Child("domain", "devices", "disks").Index(idx).Child("cdrom", "bus").String(),
```
Comment (vladikr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go]: maybe you just need `field.Index(idx).Child("cdrom", "bus").String()` here? 
or `field.Child("domain", "devices", "disks", "disk").Index(idx).Child("cdrom", "bus").String()`

Comment (stu-gott) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go]: This actually caused the other error. Thanks for catching that!

**Pull request merged**